server/migrations/2020-06-30-135809_remove_mat_views/up.sql

```diff
@@ -0,0 +1,576 @@
+returns trigger language plpgsql
+as $$
+begin
+  IF (TG_OP = 'DELETE') THEN
+    delete from user_fast where id = OLD.id;
+  ELSIF (TG_OP = 'UPDATE') THEN
+    delete from user_fast where id = OLD.id;
+    insert into user_fast select * from user_view where id = NEW.id;
+    
+    -- Refresh post_fast, cause of user info changes
+    delete from post_aggregates_fast where creator_id = NEW.id;
+    insert into post_aggregates_fast select * from post_aggregates_view where creator_id = NEW.id;
+
+    delete from comment_aggregates_fast where creator_id = NEW.id;
+    insert into comment_aggregates_fast select * from comment_aggregates_view where creator_id = NEW.id;
+
+  ELSIF (TG_OP = 'INSERT') THEN
+    insert into user_fast select * from user_view where id = NEW.id;
+  END IF;
+
+  return null;
+end $$;
+
+-- Post
+
+create table post_aggregates_fast as select * from post_aggregates_view;
+alter table post_aggregates_fast add column fast_id serial primary key;
+
+create index idx_post_aggregates_fast_id on post_aggregates_fast (id);
+
+-- For the hot rank resorting
+create index idx_post_aggregates_fast_hot_rank_published on post_aggregates_fast (hot_rank desc, published desc);
+
+create view post_fast_view as 
+select
+ap.*,
+u.id as user_id,
+coalesce(pl.score, 0) as my_vote,
+(select cf.id::bool from community_follower cf where u.id = cf.user_id and cf.community_id = ap.community_id) as subscribed,
+(select pr.id::bool from post_read pr where u.id = pr.user_id and pr.post_id = ap.id) as read,
+(select ps.id::bool from post_saved ps where u.id = ps.user_id and ps.post_id = ap.id) as saved
+from user_ u
+cross join (
+  select
+  pa.*
+  from post_aggregates_fast pa
+) ap
+left join post_like pl on u.id = pl.user_id and ap.id = pl.post_id
+
+union all
+
+select 
+paf.*,
+null as user_id,
+null as my_vote,
+null as subscribed,
+null as read,
+null as saved
+from post_aggregates_fast paf
+;
+
+drop trigger refresh_post on post;
+
+create trigger refresh_post
+after insert or update or delete
+on post
+for each row
+execute procedure refresh_post();
+
+-- Sample select
+-- select id, name from post_fast_view where name like 'test_post' and user_id is null;
+-- Sample insert 
+-- insert into post(name, creator_id, community_id) values ('test_post', 2, 2);
+-- Sample delete
+-- delete from post where name like 'test_post';
+-- Sample update
+-- update post set community_id = 4  where name like 'test_post';
+create or replace function refresh_post()
+returns trigger language plpgsql
+as $$
+begin
+  IF (TG_OP = 'DELETE') THEN
+    delete from post_aggregates_fast where id = OLD.id;
+
+    -- Update community number of posts
+    update community_aggregates_fast set number_of_posts = number_of_posts - 1 where id = OLD.community_id;
+  ELSIF (TG_OP = 'UPDATE') THEN
+    delete from post_aggregates_fast where id = OLD.id;
+    insert into post_aggregates_fast select * from post_aggregates_view where id = NEW.id;
```

Comment:
This one to me feels much less worse than the other one, because its only fetching a single id from post_aggregates_view, which I just timed at `1.2 ms`.

I agree using an `update` rather than `delete` then `insert` would be much better, but all the different possible updated column variations would make this really verbose. It can be done at a later time tho.

Reply:
I feel I'm touching on the same concept in a few places so maybe this is better served on the table defs, but generally my take away is these aggregate tables are storing "static" data like a post author, or post text, which can be retrieved from the regular post table - And then the amount of update calls done in these triggers would be really small because they only need to update stuff like "total likes" or "total comments"